### PR TITLE
[Merged by Bors] - feat: forward-port leanprover-community/mathlib#18356

### PR DESCRIPTION
**Pull request by joneugster**
This is the corresponding forward porting PR to https://github.com/leanprover-community/mathlib/pull/18356 and https://github.com/leanprover-community/mathlib/pull/18491

---

* [`data.set.basic`@`b875cbb7f2aa2b4c685aaa2f99705689c95322ad`..`75608affb24b4f48699fbcd38f227827f7793771`](https://leanprover-community.github.io/mathlib-port-status/file/data/set/basic?range=b875cbb7f2aa2b4c685aaa2f99705689c95322ad..75608affb24b4f48699fbcd38f227827f7793771)
* [`data.set.image`@`2f4cdce0c2f2f3b8cd58f05d556d03b468e1eb2e`..`4550138052d0a416b700c27056d492e2ef53214e`](https://leanprover-community.github.io/mathlib-port-status/file/data/set/image?range=2f4cdce0c2f2f3b8cd58f05d556d03b468e1eb2e..4550138052d0a416b700c27056d492e2ef53214e)

~**Note:** #1616 updated `Data.Set.Image` to commit [`2f4cdce0c2f2f3b8cd58f05d556d03b468e1eb2e`](https://github.com/leanprover-community/mathlib/commit/2f4cdce0c2f2f3b8cd58f05d556d03b468e1eb2e) but used the wrong SHA for it (it used [`b875cbb7f2aa2b4c685aaa2f99705689c95322ad`](https://github.com/leanprover-community/mathlib/commit/b875cbb7f2aa2b4c685aaa2f99705689c95322ad) which is newer, but the tools don't like that), see~

* ~[`data.set.image`@`3d95492390dc90e34184b13e865f50bc67f30fbb`..`42f4cdce0c2f2f3b8cd58f05d556d03b468e1eb2e`](https://leanprover-community.github.io/mathlib-port-status/file/data/set/image?range=3d95492390dc90e34184b13e865f50bc67f30fbb..2f4cdce0c2f2f3b8cd58f05d556d03b468e1eb2e)~

- [x] depends on: #2517

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by eric-wieser** [comment]
Can you separate marking the already-ported files into a new PR? It makes things easier to review.

**Comment by eric-wieser** [comment]
> **Note:** #1616 updated `Data.Set.Image` to commit [`2f4cdce0c2f2f3b8cd58f05d556d03b468e1eb2e`](https://github.com/leanprover-community/mathlib/commit/2f4cdce0c2f2f3b8cd58f05d556d03b468e1eb2e) but used the wrong SHA for it (it used [`b875cbb7f2aa2b4c685aaa2f99705689c95322ad`](https://github.com/leanprover-community/mathlib/commit/b875cbb7f2aa2b4c685aaa2f99705689c95322ad) which is newer, but the tools don't like that), see

I don't really understand what you mean by that.

**Comment by joneugster** [comment]
> Can you separate marking the already-ported files into a new PR? It makes things easier to review.

This might not be up to date anymore. I think now it should be fairly simple to review (except that the last sync before this one used the wrong SHA).

**Comment by joneugster** [comment]
> > **Note:** #1616 updated `Data.Set.Image` to commit [`2f4cdce0c2f2f3b8cd58f05d556d03b468e1eb2e`](https://github.com/leanprover-community/mathlib/commit/2f4cdce0c2f2f3b8cd58f05d556d03b468e1eb2e) but used the wrong SHA for it (it used [`b875cbb7f2aa2b4c685aaa2f99705689c95322ad`](https://github.com/leanprover-community/mathlib/commit/b875cbb7f2aa2b4c685aaa2f99705689c95322ad) which is newer, but the tools don't like that), see
> 
> I don't really understand what you mean by that.

@YaelDillies used a fairly new SHA (5 days ago) when the mathlib3 commit he forward-ported to was much older (about a month ago). Apparently the port-status tool wants the SHAs in the file to be exactly the one in the last sync and not any SHA equal or newer than the last sync.

**Comment by eric-wieser** [comment]
I think I understand my confusion (https://github.com/leanprover-community/mathlib4/pull/1616/files#r1118677930); @YaelDillies seems to have edited the SHA in the wrong file by accident.

**Comment by eric-wieser** [comment]
> Apparently the port-status tool wants the SHAs in the file to be exactly the one in the last sync and not any SHA equal or newer than the last sync.

The forward-port tool doesn't actually care (yet), but indeed _I_ would prefer that behavior. Note that mathlib3port now uses the desired behavior, so any new forward-ports will automatically do the right thing.

**Comment by joneugster** [comment]
> I think I understand my confusion (https://github.com/leanprover-community/mathlib4/pull/1616/files#r1118677930); @YaelDillies seems to have edited the SHA in the wrong file by accident.

You're right, I missed that these are two different files. That's definitely a typo.

**Comment by eric-wieser** [comment]
I've made https://github.com/leanprover-community/mathlib4/pull/2517 to fix just that SHA error

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#2517
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#2517~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/Data/Set/Image.lean [hunk 1]
```diff
@@ -610,6 +610,28 @@ theorem image_perm {s : Set α} {σ : Equiv.Perm α} (hs : { a : α | σ a ≠ a
 
 end Image
 
+/-! ### Lemmas about the powerset and image. -/
+
+/-- The powerset of `{a} ∪ s` is `𝒫 s` together with `{a} ∪ t` for each `t ∈ 𝒫 s`. -/
+theorem powerset_insert (s : Set α) (a : α) : 𝒫 insert a s = 𝒫 s ∪ insert a '' 𝒫 s :=
+  by
```
Comment (eric-wieser) [inline Mathlib/Data/Set/Image.lean:617]:
```suggestion
theorem powerset_insert (s : Set α) (a : α) : 𝒫 insert a s = 𝒫 s ∪ insert a '' 𝒫 s := by
```
if line length permits

**Review by eric-wieser — APPROVED** [review]
bors d+

Thanks!

**Pull request closed without merging**